src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -0,0 +1,3106 @@
+				printf("Dir %s does not exist or error to query. %s\n",
+				       full_path_loc, strerror(errno));
+				rc = 1;
+			}
+		}
+	} else {
+		*is_target_path = 0;
+		*parent = NULL;
+		item_name[0] = '\0';
+		return 0;
+	}
+
+	return rc;
+}
+
+static void
+remove_dot_dot(char szPath[])
+{
+	char *p_Offset_2Dots, *p_Back, *pTmp, *pMax, *pNewStr;
+	int i, nLen, nNonZero = 0;
+
+	nLen = strlen(szPath);
+
+	p_Offset_2Dots = strstr(szPath, "..");
+	if (p_Offset_2Dots == (szPath+1))	{
+		printf("Must be something wrong in path: %s\n", szPath);
+		return;
+	}
+
+	while (p_Offset_2Dots > 0)	{
+		pMax = p_Offset_2Dots + 2;
+		for (p_Back = p_Offset_2Dots-2; p_Back >= szPath; p_Back--) {
+			if (*p_Back == '/')	{
+				for (pTmp = p_Back; pTmp < pMax; pTmp++)
+					*pTmp = 0;
+				break;
+			}
+		}
+		p_Offset_2Dots = strstr(p_Offset_2Dots + 2, "..");
+		if (p_Offset_2Dots == NULL)
+			break;
+	}
+
+	pNewStr = szPath;
+	for (i = 0; i < nLen; i++)	{
+		if (szPath[i])	{
+			pNewStr[nNonZero] = szPath[i];
+			nNonZero++;
+		}
+	}
+	pNewStr[nNonZero] = 0;
+}
+
+static int
+remove_dot(char szPath[])
+{
+	char *p_Offset_Dots, *p_Offset_Slash, *pNewStr;
+	int i, nLen, nNonZero = 0;
+
+	nLen = strlen(szPath);
+
+	p_Offset_Dots = strstr(szPath, "/./");
+
+	while (p_Offset_Dots > 0)	{
+		p_Offset_Dots[0] = 0;
+		p_Offset_Dots[1] = 0;
+		p_Offset_Dots = strstr(p_Offset_Dots + 2, "/./");
+		if (p_Offset_Dots == NULL)
+			break;
+	}
+
+	/* replace "//" with "/" */
+	p_Offset_Slash = strstr(szPath, "//");
+	while (p_Offset_Slash > 0)	{
+		p_Offset_Slash[0] = 0;
+		p_Offset_Slash = strstr(p_Offset_Slash + 1, "//");
+		if (p_Offset_Slash == NULL)
+			break;
+	}
+
+	pNewStr = szPath;
+	for (i = 0; i < nLen; i++)	{
+		if (szPath[i])	{
+			pNewStr[nNonZero] = szPath[i];
+			nNonZero++;
+		}
+	}
+	/* remove "/" at the end of path */
+	pNewStr[nNonZero] = 0;
+	for (i = nNonZero-1; i >= 0; i--)	{
+		if (pNewStr[i] == '/')	{
+			pNewStr[i] = 0;
+			nNonZero--;
+		} else {
+			break;
+		}
+	}
+
+	return nNonZero;
+}
+
+static void
+init_fd_list(void)
+{
+	int i;
+
+	if (pthread_mutex_init(&lock_fd, NULL) != 0) {
+		printf("\n mutex create_new_lock lock_fd init failed\n");
+		exit(1);
+	}
+	if (pthread_mutex_init(&lock_dirfd, NULL) != 0) {
+		printf("\n mutex create_new_lock lock_dirfd init failed\n");
+		exit(1);
+	}
+
+	file_list = (struct FILESTATUS *)malloc(sizeof(struct FILESTATUS)*MAX_OPENED_FILE);
+	dir_list = (struct DIRSTATUS *)malloc(sizeof(struct DIRSTATUS)*MAX_OPENED_DIR);
+	memset(file_list, 0, sizeof(struct FILESTATUS)*MAX_OPENED_FILE);
+	memset(dir_list, 0, sizeof(struct DIRSTATUS)*MAX_OPENED_DIR);
+
+	for (i = 0; i < MAX_OPENED_FILE; i++)
+		file_list[i].file_obj = NULL;
+	for (i = 0; i < MAX_OPENED_DIR; i++) {
+		dir_list[i].fd = -1;
+		dir_list[i].dir_obj = NULL;
+	}
+	next_free_fd = 0;
+	last_fd = -1;
+	next_free_dirfd = 0;
+	last_dirfd = -1;
+	num_fd = num_dirfd = 0;
+}
+
+static int
+find_next_available_fd(void)
```

Comment:
I thought using a fake fd instead of the fd allocated by kernel can avoid the overhead from user space and kernel space context switching. What is the benefit to use an fd allocated by kernel (fcntl(fd, F_DUPFD, fd + FD_FILE_BASE)?